## Aritmética de punto flotante

¿Qué está pasando?

In [28]:
0.2 + 0.3

0.5

In [29]:
x = 0.3
print(x)

0.3


In [30]:
0.3 - 0.2

0.09999999999999998

- ¿Cuánto da $(\sqrt{2})^2 - 2$?


In [31]:
import numpy as np
np.sqrt(2)**2-2 

4.440892098500626e-16

- ¿Cuál es el límite cuando $n \rightarrow \infty$ de esta sucesión?
$$
\begin{aligned}
x_1 &= \sqrt{2} \\
x_{n+1} &= \frac{x_n \cdot x_n}{\sqrt{2}}
\end{aligned}
$$

In [32]:
x = np.sqrt(2)
print(x)
for i in range(100):
    x = (x * x) / np.sqrt(2)
print(x)

1.4142135623730951
inf


/tmp/ipykernel_220185/2609735010.py:4: RuntimeWarning: overflow encountered in double_scalars
  x = (x * x) / np.sqrt(2)


## Acumulación de errores

1. Se quiere calcular 
$$
\sum_{i=1}^{10^7} \frac{1}{i} \quad y \quad \sum_{i=1}^{2\cdot10^7} \frac{1}{i} 
$$
usando aritmética de simple precisión (float32).

Realizar para cada una de las expresiones un script que calcule el resultado. Qué se observa?

In [49]:
import numpy as np

n = 7
s1 = np.float32(0)
for i in range(1,10**n+1):
    s1 = s1 + np.float32(1/i)

print("suma = ", s1)

s1 = np.float32(0)
for i in range(1,2*10**n+1):
    s1 = s1 + np.float32(1/i)

print("suma2 = ", s1)


suma =  15.403683
suma2 =  15.403683


¿Qué modificación harías para reducir los errores numéricos?

In [50]:
n = 7
s1 = np.float32(0)
for i in range(1,10**n+1):
    s1 = s1 + np.float64(1/i)

print("suma = ", s1)

s1 = np.float32(0)
for i in range(1,2*10**n+1):
    s1 = s1 + np.float64(1/i)

print("suma2 = ", s1)


suma =  16.695311365857272
suma2 =  17.388458521417103


2. Utilizar las mismas estrategias para estimar $e$ mediante la serie
$$
e \approx \sum_{n=0}^{10} \frac{1}{n!}.
$$

Comparar con el valor real.

In [51]:
print(np.exp(1))

s = np.float32(0)

def factorial(n):
    f = 1
    for i in range(2,n+1,1):
        f = f * i
    return f

for i in range(1,11):
    s = s + np.float32(1/factorial(i))


print('sumatoria =', s)

2.718281828459045
sumatoria = 1.7182816


e = 2.718281828459045
sumatoria = 2.7182817

=> comparten los primeros 6 decimales

3. El siguiente código suma 1 10^8 veces. ¿Coincide la respuesta con el valor esperado? ¿Es posible modificar el código para calcular el valor correcto?

In [46]:
print("10^8 = ",10**8)

c = np.float32(0)

for i in range(10**8):
    c = c + np.float32(1)

print("10^8 en 32 = ",c)

c = np.float32(0)

for i in range(10**8):
    c = c + np.float64(1)

print("10^8 en 64 = ", c)


10^8 =  100000000
10^8 en 32 =  16777216.0
10^8 en 64 =  100000000.0
